src/pybind/mgr/dashboard/models_proposal.md

```diff
@@ -0,0 +1,608 @@
+
+* `String`: for specifying string attributes
+* `Int`: for specifying integer attributes
+* `ListOf`: for specifying a list of attributes of the same type
+* `Model`: for specifying a model class as an attribute
+
+Many other types of attributes can be built, even custom attributes. Each attribute class will
+always have at least the following four parameters in the constructor (inherited from the
+`Attribute` class):
+
+* `description: str = None`: the description of the attribute (mostly for documentation purposes).
+* `validator: Validator = None`: a validator class instance that implements the validation rule.
+* `required: bool = True`: whether the attribute is required to have a value or not.
+* `label: str = None`: the label to use when serializing the attribute in the transport format.
+
+Validator classes are used to implement validation rules for the attributes of models. Each
+validator class will extend from the `Validator` base class.
+We already provide a set of standard validator classes, but it will be easy to create custom
+validators that can be re-used across the code base.
+
+The current list of validator classes are:
+
+* `Regex`: validates if the attribute value matches regex.
+* `Length`: validates if the attribute value length is within some range.
+* `NotEmpty`: validates if the attribute value is not empty.
+* `IPAddress`: validates if the attribute value is an IP address.
+* `Enum`: validates if the attribute value is matches one of the possible options.
+* `Gt`: validates if the attribute value is greater than some lower bound.
+
+
+## How to write a model?
+
+Let's use the NFS-Ganesha controller example to specify a model that describes the requests and
+responses structure.
+
+When creating an NFS export using the CephFS backend, the dashboard frontend makes a POST request
+to the backend with the following JSON object in the request body.
+
+```json
+{
+  "cluster_id": "hello",
+  "daemons": ["node1", "node2"],
+  "fsal":{
+    "name": "CEPH",
+    "user_id": "fs_a",
+    "fs_name": "a",
+    "sec_label_xattr": "security.selinux"
+  },
+  "path": "/mydir",
+  "tag": "mytag",
+  "pseudo": "/cephfs/mydir",
+  "access_type": "RW",
+  "squash": "no_root_squash",
+  "clients": [{
+    "addresses": ["192.168.100.0/24"],
+    "access_type": "RO",
+    "squash": "no_root_squash"
+  },{
+    "addresses": ["192.168.1.103", "192.168.1.104"],
+    "access_type": "RW",
+    "squash": "all_squash"
+  }],
+  "security_label": true,
+  "protocols": [3, 4],
+  "transports": ["TCP", "UDP"]
+}
+```
+
+And the response body is almost similar to the request with exception of the addition of the
+`export_id` attribute:
+
+```json
+{
+  "export_id": 1,
+  // ...
+}
+```
+
+Let's first specify the model for the FSAL object:
+
+```python
+from .model import BodyModel, validator as Val, attribute as Attr
+
+
+class FsalModel(BodyModel):
+  name = Attr.String(
+    description="Export path",
+    validator=Val.Enum("CEPH", "RGW"))
```

Comment:
We are already using OpenAPI. Our swagger-ui page consumes an OpenAPI specification to generate the forms to try the REST API. And we are already automatically generating that OpenAPI specification. Therefore, yes, the objective is that this proposal should be aligned with the OpenAPI specification so that we can easily generate the specification.

I haven't yet gone through the OpenAPI documentation to check if there are incompatibilities with what I wrote in this proposal, but I intend to that after we reach an agreement that this is the way to go.

Reply:
Great thanks. Probably you won't need a thorough reading of that, just the notions/syntax/semantics. IMHO it doesn't make sense to have a specification if it's not somehow enforced: by generating AND consuming that specification, we are ensuring we fully comply with our spec, and we'll avoid lots of code duplication (models in front-end and back-end which sometimes which are sometimes mismatching).